API/models/model.go

```diff
@@ -652,133 +614,175 @@ func DeleteEntity(entity string, id string, userRoles map[string]Role) *u.Error
-	var updatedDoc map[string]interface{}
-	retDoc := options.ReturnDocument(options.After)
-	entInt := u.EntityStrToInt(ent)
+func prepareUpdateObject(ctx mongo.SessionContext, entity int, id string, updateData, oldObject map[string]any, userRoles map[string]Role) *u.Error {
+	// Check user permissions in case domain is being updated
+	if entity != u.DOMAIN && u.IsEntityHierarchical(entity) && (oldObject["domain"] != updateData["domain"]) {
+		if perm := CheckUserPermissions(userRoles, entity, updateData["domain"].(string)); perm < WRITE {
+			return &u.Error{Type: u.ErrUnauthorized, Message: "User does not have permission to change this object"}
+		}
+	}
 
-	//Update timestamp requires first obj retrieval
-	//there isn't any way for mongoDB to make a field
-	//immutable in a document
-	oldObj, err := GetEntity(req, ent, u.RequestFilters{}, userRoles)
+	// tag list edition support
+	err := addAndRemoveFromTags(ctx, entity, id, updateData)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	//Check if permission is only readonly
-	if entInt != u.BLDGTMPL && entInt != u.ROOMTMPL && entInt != u.OBJTMPL &&
-		(oldObj["description"] == nil) {
-		// Description is always present, unless GetEntity was called with readonly permission
-		return nil, &u.Error{Type: u.ErrUnauthorized,
-			Message: "User does not have permission to change this object"}
+	// Ensure the update is valid
+	err = ValidateEntity(entity, updateData)
+	if err != nil {
+		return err
 	}
 
-	// Update old object data with patch data
-	if isPatch {
-		var formattedOldObj map[string]interface{}
-		// Convert primitive.A and similar types
-		bytes, _ := json.Marshal(oldObj)
-		json.Unmarshal(bytes, &formattedOldObj)
-		// Update old with new
-		err := updateOldObjWithPatch(formattedOldObj, t)
+	updateData["lastUpdated"] = primitive.NewDateTimeFromTime(time.Now())
+	updateData["createdDate"] = oldObject["createdDate"]
+	delete(updateData, "parentId")
+
+	if entity == u.TAG {
+		// tag slug edition support
+		if updateData["slug"].(string) != oldObject["slug"].(string) {
+			err := repository.UpdateTagSlugInEntities(ctx, oldObject["slug"].(string), updateData["slug"].(string))
+			if err != nil {
+				return err
+			}
+		}
+
+		err := updateTagImage(ctx, oldObject, updateData)
 		if err != nil {
-			return nil, &u.Error{Type: u.ErrBadFormat, Message: err.Error()}
+			return err
 		}
-		t = formattedOldObj
-		// Remove API set fields
-		delete(t, "id")
-		delete(t, "lastUpdated")
-		delete(t, "createdDate")
 	}
 
-	// Ensure the update is valid
-	ctx, cancel := u.Connect()
-	if ok, err := ValidateEntity(u.EntityStrToInt(ent), t); !ok {
-		return nil, err
-	}
+	return nil
+}
 
-	t["lastUpdated"] = primitive.NewDateTimeFromTime(time.Now())
-	t["createdDate"] = oldObj["createdDate"]
-	delete(t, "parentId")
+func UpdateObject(entityStr string, id string, updateData map[string]interface{}, isPatch bool, userRoles map[string]Role) (map[string]interface{}, *u.Error) {
```

Comment:
We can add tags without creating them first, is it intended to be like this?

Reply:
No, this should not be possible. In models/tag.go:26 it checks for the existence of a tag before adding it. Have you found a way to do this?

Reply:
Using Postman, I sent a PUT and also a POST with tags:[tag that does not exist] and the object was created/modified with it

Reply:
Fixed in 6faa5fec

Reply:
ok for the tag that does not exist, but it accepts "tags" : "a string that is not an array", it does not add it to the object but it does create the object and gives a 201 when an error of format validation was expected.

Reply:
fixed in 54592f6c

Reply:
Another observation: if we send a list of tags and only one of them is not valid, the response is just "tag not found" and we don't know which tag.

Reply:
fixed in 5fa22073